### PR TITLE
Add kind to APIResource

### PR DESCRIPTION
**Pull request by brendandburns**
@caesarxuchao 

See the discussion in https://github.com/kubernetes/kubernetes/pull/18835 for context.

### CODE REVIEW

**Comment by caesarxuchao** [comment]
Thanks, SGTM, but I know @smarterclayton has concerns on if one resource can be mapped to multiple kinds. 

ref #18577 
@nikhiljindal @deads2k

**Comment by brendandburns** [comment]
@caesarxuchao I don't think that is a concern here, since the api installer call is called restful path on the API server, and so there can only be one Kind per RESTful path

**Comment by caesarxuchao** [comment]
I'm actually not convinced by Clayton's concern.

Apart from that, I'm considering if we should include the GroupVersion of the Kind in APIResource. My assumption is that the GroupVersion should always equal to that of the path (which is filled in [APIResourceList.GroupVersion](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/types.go#L350), so we don't need to add a GroupVersion field in APIResource.

To make my assumption valid, we need to add a check of `Version` at https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/api_installer.go#L134. Could you add that check? If that is successful, I'll be confident with this PR.

**Comment by brendandburns** [comment]
@caesarxuchao check added, ptal.

thanks!
--brendan

**Comment by caesarxuchao** [comment]
First time seeing this error on Jenkins:
```
Verifying ./hack/../hack/verify-swagger-spec.sh
+++ [0108 01:00:34] Building go targets for linux/amd64:
    cmd/kube-apiserver
+++ [0108 01:00:35] Placing binaries
+++ [0108 01:00:37] Building go targets for linux/amd64:
    cmd/kube-apiserver
+++ [0108 01:00:38] Placing binaries

etcd appears to already be running on this machine (17352 etcd) (or its a zombie and you need to kill its parent).


retry after you resolve this etcd error.
```

**Comment by caesarxuchao** [comment]
@k8s-bot test this please.

**Comment by caesarxuchao** [comment]
@brendandburns it's failing on my local machine as well, it seems the apiserver is not up.

**Comment by caesarxuchao** [comment]
The changes LGTM. I'll wait until tomorrow morning to let others chime in.

**Comment by brendandburns** [comment]
@caesarxuchao I had some test failures, they are now fixed.

I actually had to rip out the version check.  The reason for this is that `ObjectKinds` uses the go type as the key in the `map[reflect.Type]unversioned.ObjectVersionKind` and in this case, the go type is the internal API type which has no version, so the call to `ObjectKinds` only returns `ObjectVersionKind` objects with a version of empty string.

I'm not sure if this is really the expected behavior for `ObjectKinds` but it is certainly the way it is coded now and it is beyond the scope of this PR to fix that.

Because the go types are implicitly restricted to a single version, its not possible for `ObjectKinds` to return an `ObjectVersionKind` for multiple different versions, so we're still safe against some sort of clashing (or at least as safe as we would be if we had the version check still in place.)

I'm happy to go over this in detail with you tomorrow.

Thanks
--brendan

**Comment by deads2k** [comment]
> Apart from that, I'm considering if we should include the GroupVersion of the Kind in APIResource. My assumption is that the GroupVersion should always equal to that of the path (which is filled in APIResourceList.GroupVersion, so we don't need to add a GroupVersion field in APIResource.

The GroupVersion for the Kind does not always equal the GroupVersion of the Resource.  For instance, `openshift.group.io,v1,resource/scale` must live in the `openshift.group.io` API group.  However, for HPA to make use of it, the discover doc must indicate that the Kind is `extensions,v1beta1,scale`.

**Comment by brendandburns** [comment]
@deads2k is that just an observation or is it relevant to this PR?  It seems to me like we could add GroupVersion to the APIResource in a follow on PR if that's what is desired.

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by caesarxuchao** [comment]
@brendandburns, you are right, the version check I suggested won't work.
@deads2k, that's true. Subresources are tricky, let me take a closer look at them, I'm suspecting perhaps we don't need an entry for them in the RESTMapper.

**Comment by caesarxuchao** [comment]
@deads2k, it seems to me we don't need a RESTMapper entry for subresources in client.

So I'll LGTM this PR. We can add GroupVersion to APIResource if we find it necessary later.

**Comment by brendandburns** [comment]
thanks @caesarxuchao !

**Comment by brendandburns** [comment]
@k8s-bot unit test this please

**Comment by alex-mohr** [comment]
@k8s-bot test this

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by caesarxuchao** [comment]
@k8s-bot e2e test this please

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by brendandburns** [comment]
@k8s-bot unit test this please

File: pkg/api/unversioned/types.go [hunk 1]
```diff
@@ -339,6 +339,8 @@ type APIResource struct {
 	Name string `json:"name"`
 	// namespaced indicates if a resource is namespaced or not.
 	Namespaced bool `json:"namespaced"`
+	// kind is the kind for the resource (e.g. 'Foo' is the kind for a resource 'foo')
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: This leaks internal go type names into our API. @bgrant0607

Comment (deads2k) [inline pkg/api/unversioned/types.go]: > This leaks internal go type names into our API. @bgrant0607

`kind` isn't necessarily the same as the go type name.  This is `kind` in the `RESTMapper` sense that uses the `kind` in the JSON serialization.  Or at least that was the intent.

Comment (bgrant0607) [inline pkg/api/unversioned/types.go]: As mentioned in the issue, `kind` is already part of the API.

**Pull request merged**